### PR TITLE
[Immediate] Lazy Immediate Mode

### PR DESCRIPTION
**Pull request by zachary-kent**
This PR implements a WIP implementation of lazy compilation for Swift Immediate mode, where type and syntax errors in scripts are deferred to runtime. This emulates the behavior of dynamic languages, allowing errors in functions not executed at runtime to be completely ignored. Lazy compilation is disabled by default, and enabled by the experimental feature flag `LazyImmediate`.

### CODE REVIEW

**Comment by zachary0kent** [comment]
@swift-ci please test

**Comment by zachary0kent** [comment]
@swift-ci please test

**Comment by zachary0kent** [comment]
@swift-ci please test

**Comment by zachary0kent** [comment]
@swift-ci please test

**Comment by zachary0kent** [comment]
@swift-ci please test

File: include/swift/Immediate/SwiftMaterializationUnit.h [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+  ~SwiftJIT();
+
+  /// Get the dylib associated with the main program
+  llvm::orc::JITDylib &getMainJITDylib();
+
+  /// Register a the materialization unit `MU` with the `JITDylib``JD` and
+  /// create lazy reexports for all functions defined in the interface of `MU`
+  llvm::Error addSwift(llvm::orc::JITDylib &JD,
+                       std::unique_ptr<llvm::orc::MaterializationUnit> MU);
```
Comment (lhames) [inline include/swift/Immediate/SwiftMaterializationUnit.h:63]: In the future we should change this to construct the `MaterializationUnit` internally, but I think this works for now.

**Review by lhames — APPROVED** [review]
The `libImmediate` changes look good to me.

File: lib/Sema/TypeChecker.cpp [hunk 2]
```diff
@@ -307,7 +306,7 @@ TypeCheckSourceFileRequest::evaluate(Evaluator &eval, SourceFile *SF,
         }
       }
     }
-    if (CheckFunctionBodies) {
+    if (!Ctx.TypeCheckerOpts.DeferToRuntime) {
       typeCheckDelayedFunctions(*SF);
```
Comment (slavapestov) [inline lib/Sema/TypeChecker.cpp:311]: Maybe better to not add those functions to the delayed list at all then?

File: lib/SIL/IR/SILDeclRef.cpp [hunk 3]
```diff
@@ -734,7 +734,7 @@ bool SILDeclRef::isSetter() const {
 }
 
 AbstractFunctionDecl *SILDeclRef::getAbstractFunctionDecl() const {
-  return dyn_cast<AbstractFunctionDecl>(getDecl());
+  return dyn_cast_or_null<AbstractFunctionDecl>(getDecl());
```
Comment (slavapestov) [inline lib/SIL/IR/SILDeclRef.cpp:737]: When can this be null now?

Comment (zachary0kent) [reply lib/SIL/IR/SILDeclRef.cpp:737]: I modified `SILDeclRef::getDecl` to return `nullptr` if it doesn't wrap a `ValueDecl`

**Comment by zachary0kent** [comment]
@swift-ci please test

**Comment by zachary0kent** [comment]
@swift-ci please test

**Comment by zachary0kent** [comment]
@swift-ci please test

**Pull request merged**